rcl_logging_spdlog/src/rcl_logging_spdlog.cpp

```diff
@@ -0,0 +1,154 @@
+#include <rcutils/get_env.h>
+#include <rcutils/process.h>
+#include <rcutils/time.h>
+
+#include "spdlog/spdlog.h"
+#include "spdlog/sinks/basic_file_sink.h"
+
+#include "rcl_logging_spdlog/logging_interface.h"
+
+#define RC_LOGGING_RET_OK                          (0)
+#define RC_LOGGING_RET_ERROR                       (2)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static std::shared_ptr<spdlog::logger> g_root_logger;
+
+/* These are defined here to match the severity levels in rcl. They provide a consistent way for external logger
+    implementations to map between the incoming integer severity from ROS to the concept of DEBUG, INFO, WARN, ERROR,
+    and FATAL*/
+enum RC_LOGGING_LOG_SEVERITY
+{
+  RC_LOGGING_SEVERITY_UNSET = 0,  ///< The unset log level
+  RC_LOGGING_SEVERITY_DEBUG = 10,  ///< The debug log level
+  RC_LOGGING_SEVERITY_INFO = 20,  ///< The info log level
+  RC_LOGGING_SEVERITY_WARN = 30,  ///< The warn log level
+  RC_LOGGING_SEVERITY_ERROR = 40,  ///< The error log level
+  RC_LOGGING_SEVERITY_FATAL = 50,  ///< The fatal log level
+};
+
+static spdlog::level::level_enum map_external_log_level_to_library_level(int external_level)
+{
+  spdlog::level::level_enum level;
+
+  // map to the next highest level of severity
+  if (external_level <= RC_LOGGING_SEVERITY_DEBUG) {
+    level = spdlog::level::level_enum::debug;
+  } else if (external_level <= RC_LOGGING_SEVERITY_INFO) {
+    level = spdlog::level::level_enum::info;
+  } else if (external_level <= RC_LOGGING_SEVERITY_WARN) {
+    level = spdlog::level::level_enum::warn;
+  } else if (external_level <= RC_LOGGING_SEVERITY_ERROR) {
+    level = spdlog::level::level_enum::err;
+  } else if (external_level <= RC_LOGGING_SEVERITY_FATAL) {
+    level = spdlog::level::level_enum::critical;
+  }
+  return level;
+}
+
+rcl_logging_ret_t rcl_logging_external_initialize(const char * config_file, rcutils_allocator_t allocator)
+{
+  bool config_file_provided = (nullptr != config_file) && (config_file[0] != '\0');
+  if (config_file_provided) {
+    // TODO(clalancette): deal with the config file
+  } else {
+    // To be compatible with ROS 1, we want to construct a default filename of
+    // the form ~/.ros/log/<exe>_<pid>_<milliseconds-since-epoch>.log
+
+    // First get the home directory.
+    const char *homedir = rcutils_get_home_dir();
+    if (homedir == NULL) {
+      // We couldn't get the home directory; it is not really going to be
+      // possible to do logging properly, so get out of here without setting
+      // up logging.
+      return RC_LOGGING_RET_ERROR;
+    }
+
+    // SPDLOG doesn't automatically create the log directories, so make them
+    // by hand here.
+    char dotrosdir[512] = {0};
+    snprintf(dotrosdir, sizeof(dotrosdir), "%s/.ros", homedir);
+    if (!rcutils_mkdir(dotrosdir)) {
+      return RC_LOGGING_RET_ERROR;
+    }
+
+    char logdir[512] = {0};
+    snprintf(logdir, sizeof(logdir), "%s/log", dotrosdir);
```

Comment:
Check the return code of `snprintf` to make sure that it is `>0` and `<sizeof(logdir)`. 

You probably also want to use the snprintf function defined by rcutils for slightly better cross platform support: https://github.com/ros2/rcutils/blob/e52f2cfc1aac935bb0e6c35a7f3ad332a56d6d71/include/rcutils/snprintf.h#L57

Same comments for the usage of `snprintf` below as well.

Reply:
Good call about the use of snprintf.  I've switch to `rcutils_snprintf` everywhere and now check for errors.